packages/app/src/hooks/useNodeBodyHeight.tsx

```diff
@@ -0,0 +1,59 @@
+import { type NodeId } from '@ironclad/rivet-core';
+import { useCallback, useEffect, useMemo, useRef } from 'react';
+import { useRecoilValue } from 'recoil';
+import { nodesState } from '../state/graph';
+
+export interface NodeHeightCache {
+  get: (nodeId: NodeId) => number | undefined;
+
+  has: (nodeId: NodeId) => boolean;
+
+  set: (nodeId: NodeId, height: number | undefined) => void;
+}
+
+/**
+ * A cache of node heights. This is used when a node is unmounted and moved to the dragging list, since the node's
+ * body needs to be re-rendered in order to get its height. This cache allows us to avoid flickering when the node
+ * is first rendered in the dragging list.
+ */
+export const useNodeHeightCache = (): NodeHeightCache => {
+  const nodes = useRecoilValue(nodesState);
+  const ref = useRef<Record<string, number | undefined>>({});
+
+  const set = useCallback((nodeId: NodeId, height: number | undefined) => {
+    ref.current[nodeId] = height;
+  }, []);
+
+  const get = useCallback((nodeId: NodeId) => {
+    return ref.current[nodeId];
+  }, []);
+
+  const has = useCallback((nodeId: NodeId) => {
+    return nodeId in ref.current;
+  }, []);
+
+  return useMemo(() => {
+    const pre = ref.current;
+    ref.current = nodes.reduce((acc, next) => {
+      acc[next.id] = pre[next.id];
+      return acc;
+    }, {} as Record<string, number | undefined>);
+    return { set, get, has } satisfies NodeHeightCache;
+  }, [nodes, set, get, has]);
+};
+
+/**
+ * This hook persist the last known height of a node's body to the height cache, and can later use that last known
+ * height temporarily while the node is waiting for the body to be available.
+ */
+export const useNodeBodyHeight = (cache: NodeHeightCache, nodeId: NodeId, ready: boolean) => {
+  const ref = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    if (ready) {
+      cache.set(nodeId, ref.current?.getBoundingClientRect().height);
+    }
+  }, [cache, nodeId, ready]);
```

Comment:
It looks like this will call `getBoundingClientRect` for every node every time any node changes (because `nodes` will invalidate `cache` which will invalidate this effect). It's probably not a huge deal, but I have noticed that `getBoundingClientRect` can be pretty slow and lag Rivet. Is it possible to only call `getBoundingClientRect` for a node when only _it_ has changed?

Reply:
You are correct, To fix this, I changed the `useNodeHeightCache` so it now returns a stable `HeightCache` that does not change with nodes change. As a result, this will only call `getBoundingClientRect` once when `ready` becomes `true`.

The `nodes` dependency is now internal to the `useNodeHeightCache` where it uses the cache, and I also clean up the cache on an interval of 100 nodes being changes, instead on every change. This should give us the best performance that I can think of.